// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: indykite/knowledge/v1beta2/model.proto

package knowledgev1beta2

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on Return with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Return) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Return with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in ReturnMultiError, or nil if none found.
func (m *Return) ValidateAll() error {
	return m.validate(true)
}

func (m *Return) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if l := utf8.RuneCountInString(m.GetVariable()); l < 1 || l > 50 {
		err := ReturnValidationError{
			field:  "Variable",
			reason: "value length must be between 1 and 50 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(m.GetProperties()) > 50 {
		err := ReturnValidationError{
			field:  "Properties",
			reason: "value must contain no more than 50 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return ReturnMultiError(errors)
	}

	return nil
}

// ReturnMultiError is an error wrapping multiple validation errors returned by
// Return.ValidateAll() if the designated constraints aren't met.
type ReturnMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ReturnMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ReturnMultiError) AllErrors() []error { return m }

// ReturnValidationError is the validation error returned by Return.Validate if
// the designated constraints aren't met.
type ReturnValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ReturnValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ReturnValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ReturnValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ReturnValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ReturnValidationError) ErrorName() string { return "ReturnValidationError" }

// Error satisfies the builtin error interface
func (e ReturnValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sReturn.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ReturnValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ReturnValidationError{}
